Modify set_ticklabels() to fix counterintuitive behavior of set_ticklabels(get_ticklabels)#2246

MirandaXM opened the pull request:
Dear all,

I made some changes to set_ticklabels() to solve counterintuitive behavior of set_ticklabels(get_ticklabels) #2246.
Hope it helps, thank you~~

tacaswell commented on lib/matplotlib/axis.py:
```diff
@@ -1523,6 +1523,23 @@ def set_ticklabels(self, ticklabels, *args, **kwargs):
         ACCEPTS: sequence of strings
         """
+
+        #######################################################################
```
Please remove these lines.

tacaswell replied:
We don't really have things like this else where in the code base and it is good practice to keep the the code style consistent through out the library.

I think this will also fail the automated pep8 checker.

tacaswell commented on lib/matplotlib/axis.py:
```diff
@@ -1523,6 +1523,23 @@ def set_ticklabels(self, ticklabels, *args, **kwargs):
+        #######################################################################
+        # check whether the input is return value of get_ticklabels()
+        # get_ticklabels() return a cbook.silent_list of Text objects  
+        if isinstance(ticklabels, cbook.silent_list):
```
Using `isintance` is usually an anti-pattern.  The input is supposed to be a sequence of strings, so we can assume we can iterate over it.  I think something like

```python
new_labels = []
for t in tick_labels:
    try:
        new_labels.append(t.get_text())
    except AttributeError:
       new_labels.append(t)

ticklabels = new_labels
```

Would be more robust and pythonic.

tacaswell replied:
An anti-pattern is a pattern that shows up frequently in code that tends to indicate that you are doing things the hard way or in some way that is cutting against the grain of the library.   In this case, python is duck-typed (meaning that _anything_ that has the required attributes/functions can be passed in).  Explicitly checking types and changing code path based on that can produce brittle code.  For example if a user does some processing on the list returned from `get_ticklabels`, ends up with a list and then passes that list into `set_ticklabels`.  I think the list would fail this test and they would get the old behavior instead of your (correct) behavior.

tacaswell commented:
Thank you!  I left two comments (one cosmetic, one code-related).

Was there supposed to be a test to go with that image?

tacaswell commented:
For reference see http://matplotlib.org/devel/testing.html for an explaination of how to add image tests.  For this, we probably only need a png test.  Add `extensions=['png']` as a kwarg to the decorator (that is something we need to document better).

tacaswell commented:
If you are interested, I am happy to work with you to make these changes and get this fix merged.

Don't be discouraged, my first (many) PRs to matplotlib it felt like I re-did every line multiple times.

MirandaXM commented:
Hi Thomas,

Thanks for all your great help and tolerance *^. ^* I'm very glad to work with you to fix this bug~~

I have improved the fix solution based on your advice, and I'm working on the test. I will commit them as soon as possible.

Thank you very much!

MirandaXM commented:
Hi Thomas,

Thanks a lot for your detailed explanation~~ I have made some changes to the code and add corresponding test, please check it. Thank you very much!

tacaswell commented:
This looks great!

One last kind of tricky request, can you use git re-base to re-write history so that the first png was never checked in?  Probably the easiest way to do this is to squash everything to a single commit (and then use force-push to push it back to your branch on gh).

The reason to do this is that the repo is already very large and we are doing our best to keep the size from growing more than needed.

MirandaXM commented:
Hi Thomas,

I have rewritten the commit history of the issue branch, please check it~~ 

And thanks for telling me such useful way to do this *^3^* if there are any further improvement I can do, please let me know, thank you very much!

tacaswell commented:
Awesome! 

One more tiny thing (sorry if it seems like I am dragging this out, I just don't always notice everything the first time through).  The docstring should be updated to indicate that it now takes a sequence of string or `Text` objects.

MirandaXM commented:
Hi Thomas,

Thanks a lot for pointing out this issue, I will review my fix work carefully next time~~ 
I fixed the docstring, and also rewrote the history, so it is still one commit. 
If there are any further improvement I can do, please let me know, thank you very much!

tacaswell commented:
@MirandaXM Thank you and congratulations on your first MPL contribution!  Hopefully this will be the first of many.

And don't worry about missing things, this is _why_ we do code review.

MirandaXM commented:
Hi Thomas,

Much appreciate for all your great help~~ It is my pleasure that to do my 1st MPL contribution with your help, it is definitely a great lesson I have taken. I will try to do more and better in future~~

Best regards,
Miranda
